### PR TITLE
Add aws-plcuster[-aarch64] stacks 

### PR DESCRIPTION
**Pull request by stephenmsachs**
These stacks build packages defined in
https://github.com/spack/spack-configs/tree/main/AWS/parallelcluster

They use a custom container from https://github.com/spack/gitlab-runners which
includes necessary ParallelCluster software to link and build as well as an
upstream spack installation with current GCC and dependencies.

Intel and ARM software is installed and used during the build stage but removed
from the buildcache before the signing stage.

Files `configs/linux/{arch}/ci.yaml` select the necessary providers in order to
build for specific architectures (icelake, skylake, neoverse_{n,v}1).

### CODE REVIEW

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster/spack.yaml
Comment (stephenmsachs) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster/spack.yaml]: I have checked that the `rebuild-index` job will suceed after deleting a few packages in the `signing` job. But wil there be consequences when downloading?

**Review by scottwittenburg — CHANGES_REQUESTED** [review]
Looking at this, I'm noticing that you've created 4 new pipelines, associated with only 2 new stacks.  Can it be re-organized to add one pipeline (generate/build pair) per stack?  We try to avoid allowing multiple jobs to write binaries for the same hash in parallel, as the race conditions that can result often cause checksum mismatches between the generated binaries and associated meta-data files.  

But let me know if I've just missed something.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Since this definition just redefines the same image as `.aws-pcluster-generate-image`, maybe it's not needed, and instead you could extend that definition where ever you extend this one?

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Possibly copy-pasted from above?  Either way, I don't think it should be defined again here.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
Comment (stephenmsachs) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Yes, that was remainder of using different containers for different architectures. Removing.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml
Comment (stephenmsachs) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Yes, removing.

**Comment by stephenmsachs** [comment]
> Looking at this, I'm noticing that you've created 4 new pipelines, associated with only 2 new stacks. Can it be re-organized to add one pipeline (generate/build pair) per stack? We try to avoid allowing multiple jobs to write binaries for the same hash in parallel, as the race conditions that can result often cause checksum mismatches between the generated binaries and associated meta-data files.
> 
> But let me know if I've just missed something.

Thnaks for the review. I made one stack per pipeline per your suggestions. I guess there would be potential for race condistions otherwise as some of the build targets overlap.

**Comment by stephenmsachs** [comment]
@spackbot re-run pipeline

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml
Comment (stephenmsachs) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml]: I will change the mirror names to match the pipelines in another commit after the pipelines ran though.

File: share/spack/gitlab/cloud_pipelines/configs/linux/icelake/ci.yaml
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/configs/linux/icelake/ci.yaml:11]: It seems to me like `icelake` and `skylake` could be reduced to `x86_64_v4`.

The stack `spack.yaml`s can specifiy the `SPACK_TARGET_ARCH` under it's own variables.

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml:10]: This doesn't seem consistent with the icelake configs above. Am I missing something?

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml:37]: It seems like this part could be moved into `share/spack/gitlab/cloud_pipelines/configs/pcluster/ci.yaml` and included or added to the generate jobs configs.

Comment (kwryankrattiger) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml:37]: Here https://github.com/spack/spack/pull/37627/files#diff-70f9e236c9b2eff97832cfe086f0ba52c7d7fad02d03bf0c04189bc8dd95b1f0R107

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml:5]: Since this stuff is identical between the stacks, but may be worth while to merge the aarch64 stacks and the x86_64_v4 and use a matrix spec and submappings to assign the runner tags. appropriately.

```yaml
ci:
  pipeline-gen:
  - match_behavior: first
    submapping
     - match:
       - target=neoverse_n1
         build_job:
           tags: ["graviton2"]
     - match:
       - target=neoverse_v1
         build_job:
           tags: ["graviton3"]
```

I am not sure if both `neoverse_n1` and `neoverse_v1` can be concretized on gravitron3 runners, but I am pretty sure icelake and skylake can be so I would think the same backwards compatibility might apply to ARM architectures.

**Review by kwryankrattiger — COMMENTED** [review]
Some comments. Due to the deadline maybe some of these are a follow-on

File: share/spack/gitlab/cloud_pipelines/configs/linux/icelake/ci.yaml
Comment (stephenmsachs) [reply share/spack/gitlab/cloud_pipelines/configs/linux/icelake/ci.yaml:11]: You are right. I did not have a separate stack `spack.yaml` when I made these files.

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml
Comment (stephenmsachs) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml:10]: The architecture used is defined by the cluster head node. But clusters may have different architecture compute nodes, so I want the compilers to be as broad as possible. If I made this `_v4` and added a `zen2` compute node, I would not be able to use the compiler for the compute node.

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml
Comment (stephenmsachs) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml:5]: Let me take a look at this after the deadline. It seems I need to learn some more syntax.

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml
Comment (stephenmsachs) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml:37]: Yes, I can do that.

**Comment by scottwittenburg** [comment]
I wonder what's going on in this early-stage `pkgconf` [job](https://gitlab.spack.io/spack/spack/-/jobs/7020733).  It seems to find no dependencies in the buildcache and have to install them all from source.

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml:56]: Do I understand this correctly?  You are deleting from the mirror some (possibly large) subset of the binaries just built, in order to avoid re-distribution?  I thought it was going to be ok to leave these in the stack-specific mirrors, as long as we don't included them at the root.  The way it is here, isn't it going to force every matching spec to be rebuilt from source on every pipeline?

**Comment by scottwittenburg** [comment]
I'm also curious if you know whether this is something to worry about: https://gitlab.spack.io/spack/spack/-/jobs/7020733#L67

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml
Comment (stephenmsachs) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml:56]: No, I am only deleting software that's distributed as binaries from Intel/ARM anyway. This is only oneapi and armPL software. On the one hand we are on the save side for re-distribution and on the hand the installation "from source" is downloading pre-built binaries anyway.

**Comment by stephenmsachs** [comment]
> I wonder what's going on in this early-stage `pkgconf` [job](https://gitlab.spack.io/spack/spack/-/jobs/7020733). It seems to find no dependencies in the buildcache and have to install them all from source.

I am afraid it's still using the "old" container which built `gcc@12.3.0 target=neoverse_n1` by mistake. If you pull the container now it has `gcc@12.3.0 target=aarch64` as is required in the job.

**Comment by stephenmsachs** [comment]
> I'm also curious if you know whether this is something to worry about: https://gitlab.spack.io/spack/spack/-/jobs/7020733#L67

Same issue with the "old" container with wrong gcc target. I will re-run the pipeline after it completed. Not sure how long the containers are cached, but I pushed the "new" one 8h ago. I hope this is sufficient now.

**Comment by stephenmsachs** [comment]
> Looking at this, I'm noticing that you've created 4 new pipelines, associated with only 2 new stacks. Can it be re-organized to add one pipeline (generate/build pair) per stack? We try to avoid allowing multiple jobs to write binaries for the same hash in parallel, as the race conditions that can result often cause checksum mismatches between the generated binaries and associated meta-data files.
> 
> But let me know if I've just missed something.

Not sure how this works in Github but it seems the request for these changes is blocking the PR. @scottwittenburg I already pushed 1 pipelin per stack. Do I need to do anything else to get it resolved?

**Comment by scottwittenburg** [comment]
> I will re-run the pipeline after it completed.

I know you ran into a problem with spackbot causing duplicated pipelines, but normally if you push your PR branch while a pipeline is running, gitlab is supposed to notice and cancel the previous one running on the same ref.  Usually it's fairly reliable, but as @alalazo pointed out in slack, there are some gitlab bugs filed around it.  I'd go ahead and push your branch if you think a bunch of specs are going to change hashes and have to rebuild anyway.  But I'll leave the decision up to you.

**Comment by scottwittenburg** [comment]
> I already pushed 1 pipelin per stack. Do I need to do anything else to get it resolved?

Sorry I wasn't more clear about it, but I was hoping you would end up with two pipelines for two stacks, not add two more pipelines, but I get that this was easier to accomplish.  The way I think it should be done (but let me know if you agree @kwryankrattiger), is the specs which can be concretized/built on the same image, should be combined into a single stack, and then mapping rules defined in yaml configs should take care of making sure the correct arch, etc, is used for each.

But I understand it's not so clear how to achieve that, and you're under a tight deadline at this point.  So maybe you can just file an issue to clean up these new stacks/pipelines and re-examine what's going on here when time is not so tight.  Maybe in the issue, you can link to this comment for details.

Once everything builds, I'll be ok with merging this and revisiting it in a subsequent PR.

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml
Comment (kwryankrattiger) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml:10]: if that is a concern, then I don't understand why this stack is specifically building for icelake..

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml:6]: So is this stack building gcc, or does it already exist in the container?

Comment (kwryankrattiger) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml:6]: https://github.com/spack/spack/pull/37627#issuecomment-1551818251

Comment (stephenmsachs) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml:6]: There is a gcc in the container to build all the other apps.
The stack builds gcc for the buildcache to be used as compiler on the user's systems.

Comment (stephenmsachs) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml:6]: The latest pipeline picks up the correct container. There is a `gcc@12.3.0` pre-installed to build all the software.

But our ParallelCluster users will want to build software themselves. Forthat they want a recent gcc. That's the `gcc@12.3.0%gcc@7.3.1` which is made available through his pipeline. 

I hope this clarifies what I am trying to achieve.

Comment (kwryankrattiger) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-neoverse_n1/spack.yaml:6]: yup, that makes sense, thanks!

**Comment by stephenmsachs** [comment]
@scottwittenburg can we merge? I opened an issue to address merging the stacks.

**Review by scottwittenburg — APPROVED** [review]

File: share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml
Comment (scottwittenburg) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-pcluster-icelake/spack.yaml:56]: What I should have said here, but I just didn't think of at the time, is that some time ago we made it an error (see [here](https://github.com/spack/spack/blob/60b4e2128ba00798a6d5c4d0bb1e44c4230ebde6/lib/spack/spack/cmd/ci.py#L538)) for dependencies in a pipeline rebuild job to be installed from source.  The definition of "from source" there doesn't consider that it might be downloading pre-built binaries, it's either from a spack buildcache or it's "from source".

**Pull request merged**
